Autómatos
=====

[Abrir no Colab este ficheiro para o poder executar](https://colab.research.google.com/github/melroleandro-solidreturn/Matematica-Discreta-para-Hackers/blob/main/Chapter9_Automatos.ipynb)


Para além das gramáticas regulares e expressões regulares, a sintaxe de uma linguagem regular, pode ser caracterizada por autómatos finitos.

Os autómatos finitos podem ser entendidos como a idealização de um computador que tem acesso apenas a uma quantidade limitada de memória. Computadores deste tipo podem ser utilizados para controlar um elevador, as portas automáticas de um banco ou o dispositivo de injecção de um motor de combustão interna. O capítulo termina com uma secção dedicada a autómatos celulares, que estendem o conceito das máquinas de estados finitos para sistemas distribuídos. 

Nesta secção, damos preferência a representações gráficas de autómatos finitos, por meio de diagramas. Limitamos no entanto a exposição a autómatos finitos deterministas. Abaixo apresentamos o exemplo dum diagrama que representa um autómato deste tipo. Estas representações gráficas são usualmente designadas de diagramas de estados.



## Autómatos finitos

![image](images/auto1.png)


Um autómato é definido por vários estados, representado por círculos, no exemplo os estados são $q_1,q_2,q_3,q_4$. Um dos estados é identificado como estado inicial, no exemplo $q_1$. Associado a um autómato está sempre um alfabeto $\Sigma$, no exemplo $\Sigma=\{0,1\}$. Os símbolos do alfabeto são usados para rotular os arcos (setas) do autómato. Dado um estado tem de haver sempre uma transição para outro estado (possivelmente ele próprio) associada a cada elemento um dos símbolos em $\Sigma$.  No nosso exemplo, para cada estado, existe sempre duas transições, uma associada ao símbolo 0 e a outra ao símbolo 1. Todas as possíveis transições devem estar definidas, podendo um arco ter associado mais do que um símbolo.

O input para um autómato de alfabeto $\Sigma$ é uma palavra em $\Sigma^\ast$, que é aceite ou rejeitada pelo autómato. Supondo que o input, no exemplo é, $w=0110$. Inicialmente a computação tem início no estado $q_1$. O autómato lê o primeiro símbolo, que é 0. A regra de transição determina que, caso esteja em $q_1$ e se esteja a ler um 0, o estado do autómato deve continuar a ser $q_1$. Lendo depois o próximo símbolo da palavra, que é um 1. A regra de transição determina que, caso esteja em $q_1$ e se esteja a ler um 1, o estado do autómato deve passar a ser $q_2$. Lendo depois o próximo símbolo da palavra, que é outro 1. A regra de transição determina que, caso esteja em $q_2$ e se esteja a ler um 1, o estado do autómato passa a ser $q_3$. O próximo símbolo é um 0, a regra de transição determina que, o estado do autómato deve continuar a ser $q_3$. Como não existem mais símbolos na palavra para serem lidos a computação termina, no estado $q_3$.

Uma palavra é **aceite** pelo autómato caso a computação termine num estado terminal. No diagrama de estado identificamos estados terminais por um duplo ciclo. No exemplo, o autómato só tem um estado terminal $q_4$, como a computação da palavra $w=0110$ não termina num estado terminal, diz-se que $w$ não é **aceite** ou **reconhecida** pelo autómato.


Descrevemos abaixo a computação, agora da palavra $w=011011$, através duma tabela. A primeira coluna representa o estado do autómato, num dado momento, e o símbolo da coluna da direita é o símbolo que está a ser lido:

Estado | Símbolo
-------|--------
$q_1$ | 0
$q_1$ | 1
$q_2$ | 1
$q_3$ | 0
$q_3$ | 1
$q_4$ | 1
$q_5$ | Termina

A computação tem assim início no estado $q_1$, a leitura progressiva de símbolos $0,1,1,0,1$ e $1$, faz com que o autómato assuma progressivamente os estados $q_1,q_2,q_3,q_3,q_4,$ e $q_5$, terminando a computação no estado $q_5$, por falta de mais símbolos para ler. Como $q_5$ é estado terminal a palavra $011011$ diz-se reconhecida pelo autómato.

Podemos assim diz-se que, as palavras reconhecidas por um autómato definem caminhos no diagrama de estados, que partindo do estado inicial, permitem chegar a um dos estados terminais.

Devemos notar ainda que a relação de transição num autómato finito pode ser sempre descrita por uma aplicação $t$. Identificando por $E$ o conjunto de estados e por $\Sigma$ o alfabeto, a relação de transição é uma aplicação:

$$
t:E\times \Sigma\rightarrow E.
$$

No exemplo apresentado acima, esta aplicação pode ser descrita pela tabela abaixo:

$t$ | 0 | 1
------|-------|------
$q_1$ | $q_1$ | $q_2$
$q_2$ | $q_2$ | $q_3$
$q_3$ | $q_3$ | $q_4$
$q_4$ | $q_4$ | $q_4$

Que indica que, quando no estado $q_2$, por exemplo, caso esteja a ler um 0 passa para o estado $q_2$, caso esteja a ler um 1 passa para o estado $q_3$.


**Exercício 1:**

Dado o diagrama de estados que descreve um autómato finito:

![image](images/auto2.png)


Responda às seguintes questões:

1. Quais são os estados do autómato?
2. Qual é o alfabeto do autómato?
3. Qual é o estado inicial do autómato?
4. Quais são os estados terminais?
5. Qual é a sequência de estados percorrida para o input `0001011`?
6. O autómato aceita a palavra `01011`?
7. O autómato aceita a palavra vazia (λ)? E a palavra `010110`?
8. Codifique a relação de transição do autómato através de uma tabela de dupla entrada.
9. O autómato pode terminar a sua computação no estado `s₃`? Justifique.


Quando se pretende ser mais formal, representamos os autómatos através de uma estrutura algébrica:

**Definição 1:** Autómato

Um autómato $A$ é definido por um tuplo:

$$
A = (\Sigma, E, t, q, F)
$$

onde:

1. $ \Sigma$ é um alfabeto;
2. $ E $ é um conjunto, designado de **conjunto de estados**;
3. $ t $ é uma função, designada de **função de transição**:

   $$
   t : E \times \Sigma \rightarrow E
   $$

4. $ q \in E $ é um estado, designado de **estado inicial**;
5. $ F \subseteq E $ é um subconjunto de estados, designado de **estados terminais**.


**Exemplo 1:**

Seja $ A = (\Sigma, E, t, s_0, F) $ o autómato finito com:

- alfabeto $\Sigma = \{a, b\} $,
- conjunto de estados $ E = \{s_0, s_1, s_2\} $,
- estado inicial $ s_0 $,
- conjunto de estados terminais $ F = \{s_0, s_1\} $,

e cuja função de transição $ t : E \times \Sigma \rightarrow E $ é descrita pela seguinte tabela de dupla entrada:

| $ t $   | $ a $   | $ b $   |
|----------|-----------|-----------|
| $ s_0 $ | $ s_2 $ | $ s_1 $ |
| $ s_1 $ | $ s_1 $ | $ s_1 $ |
| $ s_2 $ | $ s_1 $ | $ s_2 $ |

O autómato assim definido, de forma algébrica, tem o seguinte diagrama de estados:


![image](images/auto3.png)


Usando esta notação, podemos definir formalmente:

**Definição 2:** Palavra aceite por um autómato

Seja $ A = (\Sigma, E, t, q, F) $ um autómato finito e seja $ w = w_1 w_2 \ldots w_k $ uma palavra sobre o alfabeto $ \Sigma $.

Diz-se que **o autómato $ A $ aceita a palavra $ w $** se existir uma sequência de estados $ r_1, r_2, \ldots, r_{k+1} $ tal que:

1. $ r_1 = q $;
2. $ t(r_i, w_i) = r_{i+1} $, para todo $ i = 1, \ldots, k $;
3. $ r_{k+1} \in F $.

Esta definição formaliza o facto de que a iteração da função de transição, aplicada aos símbolos da palavra a partir do estado inicial, deve conduzir o autómato a um estado terminal, de forma a que a palavra seja reconhecida.

**Definição 3:** Linguagem reconhecida por um autómato

Diz-se que o autómato finito  $A = (\Sigma, E, t, q, F) $  **reconhece a linguagem** $ L(A) $ se:

$$
L(A) = \{ w \in \Sigma^\ast \mid \text{o autómato } A \text{ aceita a palavra } w \}.
$$



**Exemplo 2:**

1. No autómato:

![image](images/auto4.png)

   O único estado terminal é $ s_0 $. As palavras que levam o estado $ s_0 $ de volta a si próprio têm de ser formadas por zero ou mais 1's consecutivos. Assim:

   $$
   L(A) = \{ 1^n : n \geq 0 \}
   $$

   que é a linguagem associada à expressão regular $1^\ast $.


2. No autómato:

![image](images/auto5.png)

   O único estado terminal é $ s_2 $. As palavras que levam o estado $ s_0 $ a $ s_2 $ são `1` e `01`. Assim:

   $$
   L(A) = \{1, 01\}
   $$

   que é a linguagem associada à expressão regular $ 1 \;|\; 01 $.



3. O autómato:

![image](images/auto6.png)

   tem como estados terminais $ s_0 $ e $s_3$. As palavras que levam $ s_0 $ a si próprio são $\lambda, 0, 00, \ldots $, ou seja, qualquer palavra formada por zero ou mais 0's consecutivos.

   As palavras que levam $ s_0 $ a $ s_3 $ são palavras do tipo: zero ou mais 0's seguidos de `10`, seguidos de qualquer sequência de 0's e 1's. Assim:

   $$
   L(A) = \{ 0^n, \; 0^n10(0|1)^\ast : n \geq 0 \}
   $$

   que é a linguagem associada à expressão regular:

   $$
   0^\ast \;|\; 0^\ast 10 (0|1)^\ast
   $$


4. O autómato:

![image](images/auto1.png)

   aceita a linguagem:

   $$
   L = \{ w \in \{0,1\}^\ast : w \text{ contém pelo menos três 1's} \}
   $$

   Note que esta linguagem é regular e está associada à expressão regular:

   $$
   0^\ast 1 0^\ast 1 0^\ast 1 (0|1)^\ast
   $$


5. O autómato:

![image](images/auto3.png)

   reconhece a linguagem regular associada à expressão:

   $$
   \lambda \;|\; b(a|b)^\ast \;|\; ab^\ast a (a|b)^\ast
   $$

**Exemplo 3:**

Pretende-se agora construir autómatos finitos que reconheçam uma linguagem dada.

1. **O conjunto de palavras binárias que começam com dois 0's**

   A palavra mais pequena nesta linguagem é `00`. Fixando $ s_0 $ como estado inicial, utiliza-se um estado auxiliar $ s_1 $, para o qual o autómato deve transitar sempre que o primeiro símbolo for `0`. Introduz-se um estado terminal $ s_2 $, atingido após a leitura de `00`.

   Após a leitura de `00`, qualquer símbolo adicional deve manter o autómato no estado terminal $ s_2 $. Palavras que não comecem com `00` devem conduzir o autómato para um estado não terminal $ s_3 $.

   
![image](images/auto7.png)

2. **O conjunto de palavras que contêm dois 0's seguidos**

   A palavra mais pequena nesta linguagem é `00`. Utilizam-se três estados $ s_0, s_1 $ e $ s_2 $, sendo $ s_0 $ o estado inicial e $ s_2 $ o estado terminal.

   - O autómato transita de $ s_0 $ para $ s_1 $ com a leitura de `0`.
   - De $ s_1 $ para $ s_2 $ com a leitura de outro `0`.
   - Se a palavra começar com uma cadeia de `1`'s, o autómato permanece em $ s_0 $, na expectativa de encontrar `00`.
   - Se, após a leitura do primeiro `0`, surgir um `1`, o autómato regressa a $ s_0 $.
   - Depois de reconhecer `00`, qualquer símbolo posterior mantém o autómato em $ s_2 $.

![image](images/auto8.png)

3. **O conjunto de palavras que **não contêm** dois 0's seguidos**

   Considerando o autómato anterior, verifica-se que:

   - Se a computação termina em $ s_0 $ ou $ s_1 $, então a palavra **não** contém a subpalavra `00`.
   - Se termina em $ s_2 $, então contém `00`.

   Assim, a linguagem das palavras que

Abaixo pode encontrar um exemplo em Python que define 4 autómatos finitos simples usando dicionários. Cada autómato é especificado com o conjunto de estados, alfabeto, estado inicial, estados finais e uma função de transição (representada como dicionário aninhado). Também incluí funções para simular o autómato e para verificar se uma palavra é aceite.

In [26]:
# Função que simula o autómato: dado um autómato (definido em dicionário) e uma palavra,
# retorna o estado final atingido.
def simulate(automata, word):
    state = automata['initial']
    for symbol in word:
        # Obtemos a transição para o símbolo atual; se não existir, retornamos None.
        state = automata['transitions'].get(state, {}).get(symbol)
        if state is None:
            return None
    return state

In [27]:
# Função que verifica se o autómato aceita a palavra:
# retorna True se o estado final estiver entre os estados finais, False caso contrário.
def accepts(automata, word):
    final_state = simulate(automata, word)
    return final_state in automata['final']

In [28]:
# -----------------------------------------------------------------
# Exemplo 1: Autómato que aceita strings que terminam com 0.
automata1 = {
    'states': {'q0', 'q1'},
    'alphabet': {'0', '1'},
    'initial': 'q0',
    'final': {'q1'},
    'transitions': {
        'q0': {'0': 'q1', '1': 'q0'},
        'q1': {'0': 'q1', '1': 'q0'}
    }
}

print("Exemplo 1: Strings que terminam com 0")
print("Input: '110' ->", accepts(automata1, "110"))  # Espera-se True (termina com 0)
print("Input: '111' ->", accepts(automata1, "111"))  # Espera-se False


Exemplo 1: Strings que terminam com 0
Input: '110' -> True
Input: '111' -> False


In [29]:
# -----------------------------------------------------------------
# Exemplo 2: Autómato que aceita strings que contêm a subpalavra "01".
automata2 = {
    'states': {'q0', 'q1', 'q2'},
    'alphabet': {'0', '1'},
    'initial': 'q0',
    'final': {'q2'},
    'transitions': {
        'q0': {'0': 'q1', '1': 'q0'},
        'q1': {'0': 'q1', '1': 'q2'},
        'q2': {'0': 'q2', '1': 'q2'}  # Estado absorvente, uma vez alcançado
    }
}

print("\nExemplo 2: Strings que contêm '01'")
print("Input: '001' ->", accepts(automata2, "001"))  # Espera-se True (contém "01")
print("Input: '111' ->", accepts(automata2, "111"))  # Espera-se False


Exemplo 2: Strings que contêm '01'
Input: '001' -> True
Input: '111' -> False


In [30]:
# -----------------------------------------------------------------
# Exemplo 3: Autómato que aceita strings que começam com 1 e terminam com 0.
# O autómato rejeita se o primeiro símbolo não for 1 e só é aceite se o último símbolo for 0.
automata3 = {
    'states': {'q0', 'q1', 'q2', 'q_dead'},
    'alphabet': {'0', '1'},
    'initial': 'q0',
    'final': {'q2'},
    'transitions': {
        'q0': {'1': 'q1', '0': 'q_dead'},
        'q1': {'1': 'q1', '0': 'q2'},
        'q2': {'1': 'q1', '0': 'q2'},
        'q_dead': {'0': 'q_dead', '1': 'q_dead'}
    }
}

print("\nExemplo 3: Strings que começam com 1 e terminam com 0")
print("Input: '1010' ->", accepts(automata3, "1010"))  # Espera-se True
print("Input: '1001' ->", accepts(automata3, "1001"))  # Espera-se False


Exemplo 3: Strings que começam com 1 e terminam com 0
Input: '1010' -> True
Input: '1001' -> False


In [31]:
# -----------------------------------------------------------------
# Exemplo 4: Autómato que aceita strings com número par de 0's.
automata4 = {
    'states': {'q_even', 'q_odd'},
    'alphabet': {'0', '1'},
    'initial': 'q_even',
    'final': {'q_even'},
    'transitions': {
        'q_even': {'0': 'q_odd', '1': 'q_even'},
        'q_odd': {'0': 'q_even', '1': 'q_odd'}
    }
}

print("\nExemplo 4: Strings com número par de 0's")
print("Input: '1010' ->", accepts(automata4, "1010"))  # Espera-se True (2 zeros)
print("Input: '101'  ->", accepts(automata4, "101"))   # Espera-se False (1 zero)


Exemplo 4: Strings com número par de 0's
Input: '1010' -> True
Input: '101'  -> False


**Exercício 2:**

Cada uma das expressões regulares abaixo define uma linguagem sobre o alfabeto $ \{a, b, c\} $.  
Para cada uma delas, apresenta um exemplo de um autómato que reconheça a linguagem correspondente.

1. $ a \;|\; b $
2. $ a^\ast \;|\; ab \;|\; ac $
3. $ a^+ \;|\; ab^\ast $
4. $ a^\ast bc^\ast b $
5. $ (a \;|\; b)^\ast c(a \;|\; b) $

Podemos demonstrar que:

**Teorema :**

Uma linguagem é regular **se, e só se**, pode ser reconhecida por um autómato finito.

Neste sentido, toda a linguagem associada a uma expressão regular, ou gerada por uma gramática regular, pode ser reconhecida por um autómato finito.

O processo de determinação da estrutura que reconhece uma determinada linguagem pode não ser trivial — especialmente quando se pretende obter diretamente um **autómato finito determinista**.

No exercício seguinte, poderá praticar com linguagens regulares de *estrutura simples*.


**Exercício 3:**

Dá exemplos de autómatos que reconheçam as seguintes linguagens regulares (em todos os casos o alfabeto é $ \{0,1\} $:

1. $ A = \{ w \in \{0,1\}^\ast : w \text{ começa por um 1 e termina com um 0} \} $
2. $ B = \{ w \in \{0,1\}^\ast : w \text{ começa por três 1's} \} $
3. $ C = \{ w \in \{0,1\}^\ast : w \text{ contém a subpalavra } 0101 \} $
4. $ D = \{ w \in \{0,1\}^\ast : \text{ ou } w \text{ começa por 0 e } |w| \text{ é ímpar, ou } w \text{ começa por 1 e } |w| \text{ é par} \} $
5. $ E = \{ w \in \{0,1\}^\ast : w \text{ não contém a subpalavra } 0101 \} $
6. $ F = \{ w \in \{0,1\}^\ast : w \text{ não começa por três 1's} \} $
7. $ G = \{ w \in \{0,1\}^\ast : |w| \leq 5 \} $
8. $ H = \{ w \in \{0,1\}^\ast : |w| > 5 \} $


## Autómatos Celulares

Os **autómatos celulares** são uma extensão do conceito de máquinas de estados finitos para sistemas distribuídos. Consistem numa grelha de células, onde cada célula tem um estado finito (por exemplo, 0 ou 1) e actualiza o seu estado com base numa regra que depende dos estados das células vizinhas. Cada célula pode ser vista como um autómato finito individual, mas o sistema como um todo opera em paralelo.

Um exemplo clássico é a **Regra 90**, um autómato celular elementar unidimensional. Aqui, cada célula tem dois vizinhos (esquerda e direita), e a regra define que o próximo estado da célula é 1 se exactamente um dos vizinhos for 1 (isto é, os pares (0,1) ou (1,0)), caso contrário, é 0. Formalmente, para uma célula:

- **Estados**: {0, 1},
- **Entrada**: estados dos vizinhos, como (0,0), (0,1), (1,0), (1,1),
- **Função de transição**: 
  - (0,1) ou (1,0) → 1,
  - (0,0) ou (1,1) → 0.

Por exemplo, numa grelha inicial [0, 1, 0], a célula central (1) tem vizinhos (0, 0), logo torna-se 0 na próxima geração. Este processo paralelo diferencia os autómatos celulares dos autómatos finitos tradicionais, que processam entradas sequencialmente.

Outro exemplo famoso é o **Jogo da Vida** de Conway, um autómato celular bidimensional onde cada célula (viva = 1, morta = 0) se actualiza com base no número de vizinhos vivos entre os oito adjacentes. Aqui, cada célula é uma máquina de estados finitos com uma regra de transição baseada na vizinhança.



Vamos implementar a **Regra 90** para autômatos celulares em um anel, seguindo sua definição formal. Cada célula tem dois vizinhos (esquerda e direita), e seu novo estadoé determinado pelo par de vizinhos, tratado como "símbolo de entrada" de uma uma máquina de estados finitos.

#### Formalização do Problema:
1. **Estados Possíveis:** `0` ou `1`.
2. **Regra de Transição:**  
   Novo estado da célula = `1` se os vizinhos são diferentes (`0,1` ou `1,0`), `0` se são iguais (`0,0` ou `1,1`).
3. **Anel de Células:**  
   - Primeira célula: vizinhos são a segunda e a última.  
   - Última célula: vizinhos são a penúltima e a primeira.  
4. **Estado Inicial:** Todas as células começam com o mesmo estado (ex: `0`).


#### Implementação em Python:

In [46]:
rule90_automaton = {
    'initial': 0,  # Estado inicial padrão
    'transitions': {
        # Para cada estado atual (0 ou 1), transições baseadas nos vizinhos (left, right)
        0: {
            (0, 0): 0,
            (0, 1): 1,
            (1, 0): 1,
            (1, 1): 0,
        },
        1: {
            (0, 0): 0,
            (0, 1): 1,
            (1, 0): 1,
            (1, 1): 0,
        },
    },
}


def simulate_step(cells,automata):
    n = len(cells)
    new_cells = []
    for i in range(n):
        current_state = cells[i]
        left = cells[(i - 1) % n]  # Vizinho esquerdo (anterior)
        right = cells[(i + 1) % n] # Vizinho direito (próximo)
        # Obter próximo estado usando o autômato (ignora current_state)
        next_state = automata['transitions'][current_state][(left, right)]
        new_cells.append(next_state)
    return new_cells

# Exemplo de simulação com 5 células e estado inicial centralizado
initial_state = 0
num_cells = 90
steps = 139
automata = rule90_automaton

# Configuração inicial: todas 0, exceto a célula do meio como 1
cells = [0] * num_cells
cells[num_cells // 2] = 1  # Célula central é 1

history = [cells.copy()]
for _ in range(steps):
    cells = simulate_step(cells,automata)
    history.append(cells.copy())

# Imprimir evolução
print("Evolução do Sistema (Regra 90):")
for step, cells in enumerate(history):
    states = ''.join('x' if num else '-' for num in cells)
    print(states)

Evolução do Sistema (Regra 90):
----------------------------------------x---------------------------------------
---------------------------------------x-x--------------------------------------
--------------------------------------x---x-------------------------------------
-------------------------------------x-x-x-x------------------------------------
------------------------------------x-------x-----------------------------------
-----------------------------------x-x-----x-x----------------------------------
----------------------------------x---x---x---x---------------------------------
---------------------------------x-x-x-x-x-x-x-x--------------------------------
--------------------------------x---------------x-------------------------------
-------------------------------x-x-------------x-x------------------------------
------------------------------x---x-----------x---x-----------------------------
-----------------------------x-x-x-x---------x-x-x-x-------------------------

## Exercícios de Revisão

**Exercício 4:**

Considere uma gramática, com símbolo inicial $ S $, com as seguintes produções:

1. $ S \rightarrow DSD $
2. $ S \rightarrow B $
3. $ B \rightarrow aCb $
4. $ B \rightarrow bCa $
5. $ C \rightarrow DCD $
6. $ C \rightarrow D $
7. $ C \rightarrow \lambda $
8. $ D \rightarrow a $
9. $ D \rightarrow b $

**Questões**

1. Dê exemplos de 3 palavras em $ L(G) $.
2. Dê exemplos de 3 palavras que não pertencem a $ L(G) $.
3. Quais das seguintes afirmações são verdadeiras ou falsas?

    - $ C \Rightarrow aba $
    - $ C \Rightarrow^\star aba $
    - $ C \Rightarrow C $
    - $ C \Rightarrow^\ast C $
    - $ DDD \Rightarrow^\ast aba $
    - $ D \Rightarrow^\ast aba $
    - $ C \Rightarrow^\star DD $

**Exercício 5:**

Para cada expressão regular abaixo, determine as produções em $ P $, de uma gramática regular (à direita) $G = (\Sigma, V, P, S) $, tal que a expressão descreva a linguagem $ L(G) $ gerada pela gramática.

1. $ a^\ast b^\ast cc $
2. $ c^\ast (a^+ | c) $
3. $ a^+ cccc $
4. $ cb^\ast c $
5. $ acb^\ast a^\ast b $
6. $ ab^\ast a $

**Exercício 6:**

Determine as linguagens regulares reconhecidas por cada um dos autómatos abaixo:

1.
   ![Autómato 1](images/auto10.png)

2. ![Autómato 2](images/auto11.png)

3. ![Autómato 3](images/auto12.png)

4. ![Autómato 4](images/auto13.png)

5. ![Autómato 5](images/auto14.png)

6. ![Autómato 6](images/auto15.png)

**Exercício 7:**

Dê exemplos de autómatos que reconheçam cada uma das seguintes linguagens regulares (em todos os casos o alfabeto é $\{0,1\}$):

1. $ A = \{ w : w \text{ o segundo símbolo é 1 e acaba com um 1} \} $
2. $ B = \{ w : w \text{ termina com três 1's} \} $
3. $ C = \{ w : w \text{ contém a subpalavra 01 ou a subpalavra 11} \} $
4. $ D = \{ w : \text{ ou } w \text{ termina em 0 se começa com 0, ou } w \text{ termina com 1 se começa com 1} \} $
5. $ E = \{ w : w \text{ não contém nem sequências 01, nem 11} \} $
6. $ H = \{ w : \text{ se } |w| > 3 \text{ termina num 1} \} $

**Exercício 8:**

Dado o diagrama de estados que descreve um autómato finito:

![Diagrama do autómato](images/auto14.png)

### Questões:

1. Quais são os estados do autómato?
2. Qual é o alfabeto do autómato?
3. Qual é o estado inicial do autómato?
4. Quais são os estados terminais do autómato?
5. Qual é a sequência de estados para o input $ 0001011 $?
6. Será que o autómato aceita a palavra $ 01011 $?
7. Será que o autómato aceita a palavra $ \lambda $ (vazia)? E a palavra $ 010110 $?
8. Codifique a relação de transição do autómato através de uma tabela de dupla entrada.
9. O autómato pode terminar a sua computação no estado $ s_1 $? Justifique.
10. Codifique o autómato através de um dicionário.
11. Implemente uma função `auto(dict, str) -> str`, tal que para um autómato descrito pelo dicionário $ G $ e para uma palavra $ w $, no alfabeto do autómato, `auto(G, w)` determina o estado final do autómato.
12. Implemente uma função `aceita(dict, set, str) -> bool`, tal que para um autómato descrito pelo dicionário \( G \), de estados terminais no conjunto $ T $ e para uma palavra $ w $, `aceita(G, T, w)` devolve `True` se e só se a palavra $ w $ é aceite pelo autómato.

## Referências

1. **"Introduction to Automata Theory, Languages, and Computation"**  
**Autores:** John E. Hopcroft, Rajeev Motwani, Jeffrey D. Ullman  
**Descrição:** Considerado a "bíblia" da teoria dos autômatos, aborda máquinas de estados finitos (DFA/NFA), linguagens regulares, e a hierarquia de Chomsky. Inclui exemplos práticos e problemas resolvidos.  
**Tópicos-chave:** Autômatos finitos, expressões regulares, máquinas de Turing, aplicações em compiladores.  


2. **"A New Kind of Science"**  
**Autor:** Stephen Wolfram  
**Descrição:** Explora autómatos celulares (como a Regra 110 e o Jogo da Vida) e sua relação com sistemas complexos. Wolfram argumenta que autômatos simples podem gerar comportamentos universais.  
**Tópicos-chave:** Regra 90, emergência, sistemas dinâmicos, computação irreducível.  


3. **"Winning Ways for Your Mathematical Plays" (Vol. 2)**  
**Autores:** Elwyn R. Berlekamp, John H. Conway, Richard K. Guy  
**Descrição:** Dedicado a jogos matemáticos, inclui uma análise detalhada do **Jogo da Vida de Conway**, suas regras e padrões (como "gliders" e "still lifes").  
**Tópicos-chave:** Jogo da Vida, autômatos celulares, padrões autorreplicantes.  


4. **"Introduction to the Theory of Computation"**  
**Autor:** Michael Sipser  
**Descrição:** Focado na interseção entre teoria da computação e linguagens formais. Explica autómatos finitos, máquinas de estados e sua aplicação em problemas de decisão.  
**Tópicos-chave:** Linguagens regulares, decidibilidade, redução de estados.  

